### PR TITLE
Allow non-browser analytics proxy requests without Origin only when authenticated via shared token headers.

### LINKED ISSUE #12308: bug: Analytics proxies reject requests that lack an Origin header (breaks non-browser/embedded clients)
### Describe the bug

isAllowedOrigin in analytics_proxy_test.go returns false (403) if the Origin header is missing. Some legitimate non-browser clients, server-side trackers, or embedded integrations may not set Origin and will be rejected.
This will break automated analytics submission from non-browser contexts.

Image

Subtest Missing_Origin passes, which confirms no origin header is rejected.

### Environment

Windows Local Env

### Steps To Reproduce

1. Send a POST to web/src/lib/api.ts or the Umami proxy endpoint with a tool like curl without an Origin header.
2. Observe a 403 Forbidden from the proxy (the code explicitly rejects requests missing Origin).

### Expected Behavior

This should not reject legitimate non-browser clients, server-side trackers.

### Browser Console / Backend Logs

```shell

```

### Want to contribute?

- [Y] I would like to work on this issue.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by lakshyajn**
> **Adding or modifying a card/dashboard?** 
No

> **New CNCF project card?** 
No

> **Use a coding agent.** 
Yes, but for testing at first stage (GPT Codex 5.3)

### 📌 Fixes

Fixes #12308  

---

### 📝 Summary of Changes

This PR allows legitimate non-browser requests (which may not send an `Origin` header) to access the analytics proxy only when authenticated with a shared server token.

- Added a server-auth fallback for missing `Origin` requests:
  - `X-KC-Client-Auth: <token>`
  - `Authorization: Bearer <token>`
- Token is validated against `ANALYTICS_PROXY_AUTH_TOKEN`.
- Comparison uses constant-time checks for safer secret validation.
- Browser-origin allowlist behavior remains unchanged.

---

### Changes Made

- [x] Updated analytics proxy request validation logic to support authenticated non-browser callers without `Origin`
- [x] Added secure token-based auth fallback using env-configured shared secret
- [x] Added/updated unit tests for missing-origin + auth/no-auth paths
- [x] Preserved existing allowed-origin and forbidden-origin test coverage

---

### Checklist

Please ensure the following before submitting your PR:

- [x] I used a coding agent (Claude Code, Copilot, Gemini, or Codex) to generate/review this code
- [x] I have reviewed the project's contribution guidelines
- [x] New cards target console-marketplace, not this repo
- [x] isDemoData is wired correctly (cards show Demo badge when using demo data) (N/A: no card/dashboard changes)
- [x] I have written unit tests for the changes (if applicable)
- [x] I have tested the changes locally and ensured they work as expected
- [x] All commits are signed with DCO (`git commit -s`)

---

### Screenshots or Logs (if applicable)

image

---

### CODE REVIEW

File: pkg/api/handlers/analytics_proxy.go [hunk 1]
```diff
@@ -232,20 +235,44 @@ func isAllowedNetlifyHost(host string) bool {
...
+		if subtle.ConstantTimeCompare([]byte(token), []byte(expected)) == 1 {
+			return true
+		}
+	}
+
+	auth := strings.TrimSpace(c.Get("Authorization"))
+	if strings.HasPrefix(auth, bearerPrefix) && len(auth) > bearerPrefixLen {
+		token := strings.TrimSpace(auth[bearerPrefixLen:])
+		return subtle.ConstantTimeCompare([]byte(token), []byte(expected)) == 1
+	}
```
[the remark this answers is not in the document]
Comment (lakshyajn) [reply pkg/api/handlers/analytics_proxy.go:257]: @copilot apply changes based on [this feedback](https://github.com/kubestellar/console/pull/12336#discussion_r3196828213)

File: pkg/api/handlers/analytics_proxy_test.go [hunk 2]
```diff
@@ -30,6 +30,8 @@ func TestIsAllowedNetlifyHost(t *testing.T) {
 }
 
 func TestIsAllowedOrigin(t *testing.T) {
+	t.Setenv(analyticsProxyAuthTokenEnv, "test-analytics-token")
+
 	app := fiber.New()
```
[the remark this answers is not in the document]
Comment (lakshyajn) [reply pkg/api/handlers/analytics_proxy_test.go:35]: @copilot apply changes based on [this feedback](https://github.com/kubestellar/console/pull/12336#discussion_r3196828247)

**Review by clubanderson — COMMENTED** [review]
## Review

The auth hardening itself is solid — `crypto/subtle.ConstantTimeCompare`, empty-token guard, no secret leakage. A few items:

### 1. `os.Getenv()` on every request
`hasValidAnalyticsServerAuth` reads `ANALYTICS_PROXY_AUTH_TOKEN` from the environment on every origin-less request. Consider caching it at init time — avoids repeated syscalls under load and prevents accidental runtime mutation.

### 2. Missing test: both headers present with conflicting values
If a request sends a valid `Authorization: Bearer` but an invalid `X-KC-Client-Auth`, the function rejects it (because `X-KC-Client-Auth` is checked first and fails). This is actually the safer behavior, but there's no test covering it. Worth adding a test case like:
```go
{"Missing Origin Valid Bearer Invalid XAuth", "", "localhost", "bad-token", "Bearer test-analytics-token", false},
```

### 3. Missing test: feature disabled (empty env var)
No test verifying that when `ANALYTICS_PROXY_AUTH_TOKEN` is unset, all origin-less requests are rejected regardless of headers. The current tests always set the env var via `t.Setenv` at the top of `TestIsAllowedOrigin`.

### 4. Context question
Issue #12308 was filed today by the same author as this PR. The current `main` branch already returns `true` for missing-origin requests (open access, merged earlier today). This PR tightens that back to require a token — which is the right direction. But what is the actual use case for non-browser analytics submission? Is there a server-side tracker or integration that needs this? Understanding the caller helps validate the auth design.

### 5. Minor: length side-channel
`subtle.ConstantTimeCompare` returns 0 immediately when slices differ in length (per Go docs). An attacker could probe token length. Low risk given this is an analytics proxy token, but normalizing to a fixed-length hash comparison would close it entirely.

Overall the security changes are clean and focused. The items above are suggestions, not blockers.

**Comment by clubanderson** [comment]
Closing — the underlying change (#12326) is being reverted in #12348. There are no server-side analytics callers in the codebase; all GA4 events come from the browser which always sends Origin. The original 403 for missing Origin was intentional security hardening.
